traffic_ops/app/lib/API/CachegroupFallback.pm

```diff
@@ -0,0 +1,283 @@
+# limitations under the License.
+#
+#
+#
+# a note about locations and cachegroups. This used to be "Location", before we had physical locations in 12M. Very confusing.
+# What used to be called a location is now called a "cache group" and location is now a physical address, not a group of caches working together.
+#
+
+# JvD Note: you always want to put Utils as the first use. Sh*t don't work if it's after the Mojo lines.
+use UI::Utils;
+use Mojo::Base 'Mojolicious::Controller';
+use Data::Dumper;
+use JSON;
+use MojoPlugins::Response;
+use Validate::Tiny ':all';
+
+sub delete {
+	my $self = shift;
+	my $cache_id = $self->param('cacheGroupId');
+	my $fallback_id = $self->param('fallbackId');
+	my $params = $self->req->json;
+	my $rs_backups = undef; 
+
+	if ( !&is_oper($self) ) {
+		return $self->forbidden();
+	}
+
+	if ( defined ($cache_id) && defined($fallback_id) ) {
+		$rs_backups = $self->db->resultset('CachegroupFallback')->search( { primary_cg => $cache_id , backup_cg => $fallback_id} );
+	} elsif (defined ($cache_id)) {
+		$rs_backups = $self->db->resultset('CachegroupFallback')->search( { primary_cg => $cache_id} );
+	} elsif (defined ($fallback_id)) {
+		$rs_backups = $self->db->resultset('CachegroupFallback')->search( { backup_cg => $fallback_id} );
+	}
+
+	if ( ($rs_backups->count > 0) ) {
+		my $del_records = $rs_backups->delete();
+		if ($del_records) {
+			&log( $self, "Backup configuration DELETED", "APICHANGE");
+			return $self->success_message("Backup configuration DELETED");
+		} else {
+			return $self->alert( "Backup configuration DELETED." );
+		}
+	} else {
+		&log( $self, "No backup Cachegroups found");
+		return $self->not_found();
+	}
+}
+
+sub show {
+	my $self = shift;
+	my $cache_id = $self->param("cacheGroupId");
+	my $fallback_id = $self->param("fallbackId");
+	my $id = $cache_id ? $cache_id : $fallback_id;
+
+	#only integers
+	if ( $id !~ /^\d+?$/ ) {
+		&log( $self, "No such Cachegroup id $id");
+		return $self->not_found();
+	}
+
+	my $cachegroup = $self->db->resultset('Cachegroup')->search( { id => $id } )->single();
+	if ( !defined($cachegroup) ) {
+		&log( $self, "No such Cachegroup $id");
+		return $self->not_found();
+	}
+
+	if ( ($cachegroup->type->name ne "EDGE_LOC") ) {
+		return $self->alert("cachegroup should be type EDGE_LOC.");
+	}
+
+	my $rs_backups = undef;
+
+	if ( defined ($cache_id) && defined ($fallback_id)) {
+		$rs_backups = $self->db->resultset('CachegroupFallback')->search({ primary_cg => $cache_id, backup_cg => $fallback_id}, {order_by => 'set_order'});
+	} elsif ( defined ($cache_id) ) {
+		$rs_backups = $self->db->resultset('CachegroupFallback')->search({ primary_cg => $cache_id}, {order_by => 'set_order'});
+	} elsif ( defined ($fallback_id) ) {
+		$rs_backups = $self->db->resultset('CachegroupFallback')->search({ backup_cg => $fallback_id}, {order_by => 'set_order'});
+	}
+
+	if ( defined ($rs_backups) && ($rs_backups->count > 0) ) {
```

Comment:
just return an empty array if count=0. don't return a 404 ($self->not_found())

Reply:
I am not sure on why we shouldn't return 404. To me it looks like a valid behavior since we cant find records for such a query.

Reply:
to ensure we have a consistent API, you need to follow the examples of the other apis. If you ask for a list of resources and there are none, an empty array is returned instead of a 404

for example, 

GET /api/foos?active=true - if there are no active foos, then [] is returned and not 404.

404 implies an invalid resource path. /api/foos?active=true is a valid resource path, there just happens to be no active foos.

this on the other  hand would result in a 404:

GET /api/foos/777/bars <-- in this case there is no foo with id=777, so 404 would be valid.

Reply:
Ok. Will address this.